libs/langchain/langchain/prompts/chat.py

```diff
@@ -592,6 +593,41 @@ def partial(self, **kwargs: Union[str, Callable[[], str]]) -> ChatPromptTemplate
         prompt_dict["partial_variables"] = {**self.partial_variables, **kwargs}
         return type(self)(**prompt_dict)
 
+    def append(self, message: MessageLikeRepresentation) -> None:
+        """Append message to the end of the chat template.
+
+        Args:
+            message: representation of a message to append.
+        """
+        self.messages.append(_convert_to_message(message))
+
+    def extend(self, messages: Sequence[MessageLikeRepresentation]) -> None:
+        """Extend the chat template with a sequence of messages."""
+        self.messages.extend([_convert_to_message(message) for message in messages])
+
+    @overload
+    def __getitem__(self, index: int) -> MessageLike:
+        ...
+
+    @overload
+    def __getitem__(self, index: slice) -> ChatPromptTemplate:
+        ...
+
+    def __getitem__(
+        self, index: Union[int, slice]
+    ) -> Union[MessageLike, ChatPromptTemplate]:
+        """Use to index into the chat template."""
+        if isinstance(index, slice):
```

Comment:
not v familiar with overloading, is this better than implementing each of the specific type signatures above?

Reply:
That's the mypy syntax for implementing type discrimination. The implementation can only be provided once 

https://docs.python.org/3/library/typing.html#typing.overload